# Sistema Agéntico de Predicción - Premier League

## Orquestador del Pipeline Agéntico

Este notebook coordina la ejecución secuencial de tres skills complementarias para análisis predictivo de partidos de la Premier League:

1. **preparacion-datos** ➡️ Carga, limpieza y enriquecimiento de datos
2. **analisis-exploratorio** ➡️ EDA con análisis descriptivos y visualizaciones
3. **modelado-predictivo** ➡️ Entrenamiento y evaluación de modelos ML

El sistema sigue el estándar de **Anthropic Agent Skills** con una estructura modular donde cada skill es independiente pero forma parte de un pipeline coherente.

---

## Celda 1: Configuración Inicial del Agente

Cargar las herramientas necesarias y configurar el entorno de ejecución.

In [ ]:
import subprocess
import os
import json
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import Image, display, Markdown
from pathlib import Path

# Configurar visualización
%matplotlib inline
plt.style.use('seaborn-v0_8-darkgrid')

# Ruta de la base de datos
db_path = r"C:\Users\juanc\Documents\Universidad\Programacion Analitica De Datos\ProyectoAgentico_BDFutbol\data\premier_league.db"

print("="*70)
print("SISTEMA AGÉNTICO DE PREDICCIÓN - PREMIER LEAGUE")
print("="*70)
print(f"\n✓ Base de datos ubicada: {os.path.exists(db_path)}")
print(f"✓ Entorno configurado")
print(f"✓ Skills disponibles: 3 (preparacion, eda, modelado)")
print("\nEstándar: Anthropic Agent Skills Framework")

✓ Setup e imports completados


## Celda 2: Registro de Skills Disponibles

Mostrar la definición de cada skill desde sus archivos SKILL.md para confirmar que el agente conoce sus capacidades.

In [ ]:
# Leer y mostrar las definiciones de skills
skills_paths = {
    'preparacion-datos': 'skills/preparacion/SKILL.md',
    'analisis-exploratorio': 'skills/eda/SKILL.md',
    'modelado-predictivo': 'skills/modelado/SKILL.md'
}

print("\n" + "="*70)
print("REGISTRO DE SKILLS DEL AGENTE")
print("="*70)

for skill_name, skill_path in skills_paths.items():
    if os.path.exists(skill_path):
        with open(skill_path, 'r', encoding='utf-8') as f:
            contenido = f.read()
            # Extraer frontmatter YAML
            import re
            match = re.search(r'---\n(.*?)\n---', contenido, re.DOTALL)
            if match:
                frontmatter = match.group(1)
                print(f"\n✓ {skill_name}:")
                for line in frontmatter.split('\n'):
                    if line.strip():
                        print(f"  {line}")
    else:
        print(f"\n⚠ {skill_name}: No encontrado")

        
print("\n" + "="*70)
print("Todas las skills están registradas y disponibles para ejecución")
print("="*70)

DB existe: True


╔════════════════════════════════════════════════════════════════════╗
║                                                                    ║
║     SISTEMA AGÉNTICO - PREDICCIÓN DE RESULTADOS PREMIER LEAGUE     ║
║                                                                    ║
╚════════════════════════════════════════════════════════════════════╝


Base de datos: C:\Users\juanc\Documents\Universidad\Programacion Analitica De Datos\ProyectoAgentico_BDFutbol\data\premier_league.db
Iniciando orquestación de skills...


██████████████████████████████████████████████████████████████████████
█ SKILL 1/3: PREPARACIÓN DE DATOS
██████████████████████████████████████████████████████████████████████

INICIANDO PIPELINE DE PREPARACIÓN DE DATOS

[1/5] Cargando datos...

[2/5] Limpiando datos...
✓ Duplicados eliminados: 0
✓ Tipos de datos convertidos correctamente
✓ Datos ordenados por fecha. Rango: 2024-08-16 a 2025-05-25

[3/5] Creando features históricas...
Calculando feat

c:\Users\juanc\Documents\Universidad\Programacion Analitica De Datos\ProyectoAgentico_BDFutbol\scripts\skills\skill_eda.py:215: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  ax = sns.countplot(


✓ Gráfica 1 guardada: graphs/01_distribucion_resultados.png
✓ Gráfica 2 guardada: graphs/02_promedio_goles_equipos.png
✓ Gráfica 3 guardada: graphs/03_heatmap_correlacion.png
----------------------------------------------------------------------

📋 GENERANDO REPORTE EDA...
----------------------------------------------------------------------

ESTADISTICAS_GENERALES:
El dataset contiene 380 partidos jugados entre 20 equipos únicos en el rango 2024-08-16 a 2025-05-25. El promedio de goles por partido es 2.93, con 1.51 goles en promedio para locales y 1.42 para visitantes.

DISTRIBUCION_RESULTADOS:
La variable objetivo 'resultado' muestra: Local: 155 (40.8%), Empate: 93 (24.5%), Visitante: 132 (34.7%). Los equipos locales tienen una tasa de victoria del 40.8%, indicando una ventaja significativa del factor local.

EQUIPOS_DOMINANTES:
El equipo más ganador como local es Liverpool con 14 victorias de 19 partidos (73.7%). Como visitante, Liverpool lidera con 11 victorias de 19 partidos (57.

TypeError: only integer scalar arrays can be converted to a scalar index

## Celda 3: Ejecutar Skill - Preparación de Datos

**Entrada**: `premier_league.db` (380 partidos)  
**Salida**: `data/dataset_preparado.csv` (con features históricas)

Esta skill es el primer paso del pipeline.

In [ ]:
print("\n" + "="*70)
print("[SKILL 1/3] EJECUTANDO: preparacion-datos")
print("="*70)

try:
    resultado = subprocess.run(
        ['python', 'skills/preparacion/preparacion.py', db_path],
        capture_output=True,
        text=True,
        timeout=300
    )
    
    print(resultado.stdout)
    
    if resultado.returncode != 0:
        print(f"❌ Error en ejecución:\n{resultado.stderr}")
    else:
        print("\n✅ SKILL 1 COMPLETADA EXITOSAMENTE")
        
except subprocess.TimeoutExpired:
    print("❌ Timeout: La ejecución tardó más de 5 minutos")
except Exception as e:
    print(f"❌ Error: {e}")


🚀 INICIANDO PIPELINE AGENTICO...

Usando base de datos: c:\Users\juanc\Documents\Universidad\Programacion Analitica De Datos\ProyectoAgentico_BDFutbol\notebooks\..\premier_league.db



╔════════════════════════════════════════════════════════════════════╗
║                                                                    ║
║     SISTEMA AGÉNTICO - PREDICCIÓN DE RESULTADOS PREMIER LEAGUE     ║
║                                                                    ║
╚════════════════════════════════════════════════════════════════════╝


Base de datos: c:\Users\juanc\Documents\Universidad\Programacion Analitica De Datos\ProyectoAgentico_BDFutbol\notebooks\..\premier_league.db
Iniciando orquestación de skills...


██████████████████████████████████████████████████████████████████████
█ SKILL 1/3: PREPARACIÓN DE DATOS
██████████████████████████████████████████████████████████████████████

INICIANDO PIPELINE DE PREPARACIÓN DE DATOS

[1/5] Cargando datos...

✗ ERROR EN SKILL 1: Execution fa

Traceback (most recent call last):
  File "c:\Users\juanc\AppData\Local\Programs\Python\Python312\Lib\site-packages\pandas\io\sql.py", line 2702, in execute
    cur.execute(sql, *args)
sqlite3.OperationalError: no such table: fixtures

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\juanc\AppData\Local\Temp\ipykernel_24560\2319468751.py", line 6, in <module>
    resultados = ejecutar_pipeline_agentico(db_path)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\juanc\Documents\Universidad\Programacion Analitica De Datos\ProyectoAgentico_BDFutbol\notebooks\..\scripts\orchestrator.py", line 85, in ejecutar_pipeline_agentico
    df_preparado = preparar_datos(db_path)
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\juanc\Documents\Universidad\Programacion Analitica De Datos\ProyectoAgentico_BDFutbol\notebooks\..\scripts\skills\skill_preparacion.py", line 373, in preparar_datos
    df = c

## Celda 4: Ejecutar Skill - Análisis Exploratorio

**Entrada**: `data/dataset_preparado.csv` (generado por skill anterior)  
**Salida**: 
- `data/reporte_eda.json` (hallazgos principales)
- `graphs/01_*.png`, `graphs/02_*.png`, `graphs/03_*.png` (visualizaciones)

Esta skill es el segundo paso del pipeline.

In [ ]:
print("\n" + "="*70)
print("[SKILL 2/3] EJECUTANDO: analisis-exploratorio")
print("="*70)

try:
    resultado = subprocess.run(
        ['python', 'skills/eda/eda.py'],
        capture_output=True,
        text=True,
        timeout=300
    )
    
    print(resultado.stdout)
    
    if resultado.returncode != 0:
        print(f"❌ Error en ejecución:\n{resultado.stderr}")
    else:
        print("\n✅ SKILL 2 COMPLETADA EXITOSAMENTE")
        
except subprocess.TimeoutExpired:
    print("❌ Timeout: La ejecución tardó más de 5 minutos")
except Exception as e:
    print(f"❌ Error: {e}")

⚠️  ADVERTENCIA: El pipeline no se ejecutó correctamente
Revisa la celda anterior para ver el error


In [ ]:
print("\n" + "="*70)
print("[SKILL 3/3] EJECUTANDO: modelado-predictivo")
print("="*70)

try:
    resultado = subprocess.run(
        ['python', 'skills/modelado/modelado.py'],
        capture_output=True,
        text=True,
        timeout=300
    )
    
    print(resultado.stdout)
    
    if resultado.returncode != 0:
        print(f"❌ Error en ejecución:\n{resultado.stderr}")
    else:
        print("\n✅ SKILL 3 COMPLETADA EXITOSAMENTE")
        
except subprocess.TimeoutExpired:
    print("❌ Timeout: La ejecución tardó más de 5 minutos")
except Exception as e:
    print(f"❌ Error: {e}")

📈 ESTADÍSTICAS DESCRIPTIVAS


NameError: name 'df_preparado' is not defined

## Celda 5: Ejecutar Skill - Modelado Predictivo

**Entrada**: `data/dataset_preparado.csv` (dataset preparado)  
**Salida**:
- `data/reporte_modelado.json` (métricas y resultados)
- `graphs/04_*.png`, `graphs/05_*.png` (visualizaciones de desempeño)

Esta skill es el tercer y final paso del pipeline.

In [ ]:
## Celda 6: Resultados Finales del Modelado

Mostrar el reporte JSON con métricas y recomendaciones del modelo ganador.

In [ ]:
print("\n" + "="*70)
print("RESULTADOS FINALES DEL SISTEMA AGÉNTICO")
print("="*70)

# Cargar reporte de modelado
if os.path.exists('data/reporte_modelado.json'):
    with open('data/reporte_modelado.json', 'r', encoding='utf-8') as f:
        reporte = json.load(f)
    
    print(f"\n🏆 MODELO GANADOR: {reporte['mejor_modelo']}")
    print("-"*70)
    
    print(f"\n📊 MÉTRICAS DE DESEMPEÑO:")
    print(f"  - Accuracy:  {reporte['metricas']['accuracy']:.4f} ({reporte['metricas']['accuracy']*100:.2f}%)")
    print(f"  - Precision: {reporte['metricas']['precision']:.4f}")
    print(f"  - Recall:    {reporte['metricas']['recall']:.4f}")
    
    print(f"\n📋 ANÁLISIS DETALLADO:")
    for clave, contenido in reporte['reporte_modelado'].items():
        print(f"\n{clave.upper().replace('_', ' ')}:")
        print(f"{contenido}")
        
    print("\n" + "="*70)
else:
    print("\n⚠ Reporte de modelado no encontrado. Verificar ejecución anterior.")

## Celda 7: Visualizaciones Generadas

Mostrar todas las gráficas generadas por las skills del análisis exploratorio y modelado.

In [ ]:
# Mostrar gráficas del EDA
print("\n" + "="*70)
print("VISUALIZACIONES - ANÁLISIS EXPLORATORIO")
print("="*70)

grafica_eda = [
    'graphs/01_distribucion_resultados.png',
    'graphs/02_promedio_goles_equipos.png',
    'graphs/03_heatmap_correlacion.png'
]

for archivo in grafica_eda:
    if os.path.exists(archivo):
        print(f"\n✓ {archivo}")
        display(Image(filename=archivo))
    else:
        print(f"\n⚠ {archivo} no encontrado")

        
# Mostrar gráficas de modelado
print("\n" + "="*70)
print("VISUALIZACIONES - MODELADO PREDICTIVO")
print("="*70)

grafica_modelado = [
    'graphs/04_matriz_confusion_logisticregression.png',
    'graphs/04_matriz_confusion_randomforestclassifier.png',
    'graphs/05_comparacion_modelos.png'
]

for archivo in grafica_modelado:
    if os.path.exists(archivo):
        print(f"\n✓ {archivo}")
        display(Image(filename=archivo))
    else:
        print(f"\n⚠ {archivo} no encontrado")